crates/katana/core/src/backend/gas_oracle.rs

```diff
@@ -1,25 +1,282 @@
+use std::backtrace::Backtrace;
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::future::IntoFuture;
+use std::pin::Pin;
+
+use alloy_provider::{Provider, ProviderBuilder};
+use alloy_rpc_types_eth::BlockNumberOrTag;
+use alloy_transport::Transport;
+use anyhow::{Context, Ok};
+use futures::Future;
 use katana_primitives::block::GasPrices;
+use tokio::time::Duration;
+use url::Url;
+
+const BUFFER_SIZE: usize = 60;
+const INTERVAL: Duration = Duration::from_secs(60);
+const ONE_GWEI: u128 = 1_000_000_000;
 
 // TODO: implement a proper gas oracle function - sample the l1 gas and data gas prices
 // currently this just return the hardcoded value set from the cli or if not set, the default value.
 #[derive(Debug)]
-pub struct L1GasOracle {
+pub enum L1GasOracle {
+    Fixed(FixedL1GasOracle),
+    Sampled(SampledL1GasOracle),
+}
+
+#[derive(Debug)]
+pub struct FixedL1GasOracle {
     gas_prices: GasPrices,
     data_gas_prices: GasPrices,
 }
 
+#[derive(Debug, Default, Clone)]
+pub struct SampledL1GasOracle {
+    gas_prices: GasPrices,
+    data_gas_prices: GasPrices,
+}
+
+#[derive(Debug, Clone)]
+pub struct GasOracleWorker {
+    pub l1_oracle: SampledL1GasOracle,
+    pub l1_provider_url: Option<Url>,
```

Comment:
```suggestion
    pub l1_provider_url: Url,
```

Reply:
Hi @kariy Thanks for the review! I’ll address the changes shortly.